### PR TITLE
Materialize singleton aggregate

### PR DESCRIPTION
**Pull request by rdowavic**
This transformer replaces inline singleton aggregates with a variable.
By singleton, I mean that the aggregate, when imagined as a relation, only ever takes one value. This is so the aggregate value can be reused.
HOWEVER, as of now, there is no reuse! It will create two relations for each aggregate, even if the aggregates are identical. I am not yet sure how to resolve this. 
(how to check two aggregates are semantically equal, I am not sure?)
Here is a small example:
```
C(x) :- A(x), B(x), x < sum y : B(y).
```
The inline aggregate here doesn't depend on groundings from the outer scope, so its value will always be the same.

This becomes:
```
C(x) :- A(x), B(x), x < z, A'(z).
A'(z) :- z = sum y : { B(y) }.
```

### CODE REVIEW

**Comment by mmcgr** [comment]
You can compare two AstNodes with '=='. If that doesn't work, then there may be a bug in our tests of equality in `src/test/ast_program_test.cpp`

**Comment by rdowavic** [comment]
> You can compare two AstNodes with '=='. If that doesn't work, then there may be a bug in our tests of equality in `src/test/ast_program_test.cpp`

Okay, I will try to get something done with this. It might take me a while, but it would be really helpful and make the transformation more worthwhile.

**Comment by rdowavic** [comment]
Is it ok if we still merge this pull request then I can do a separate one for the improvement?

**Comment by mmcgr** [comment]
> Is it ok if we still merge this pull request then I can do a separate one for the improvement?

Yes, that would be preferable.

File: src/C.csv [hunk 1]
```diff
@@ -0,0 +1,4 @@
+2	1
```
Comment (mmcgr) [inline src/C.csv:1]: Remove this file

File: src/AstTransforms.cpp [hunk 2]
```diff
@@ -221,6 +222,125 @@ bool UniqueAggregationVariablesTransformer::transform(AstTranslationUnit& transl
...
+            }
+        };
+        replaceAggregate update(*aggregate, std::move(variable));
+        clause->apply(update);
+        clause->addToBody(std::move(aggHead));
+    }
+    return pairs.size() > 0;
+}
+
+// An aggregate is single-valued if
```
Comment (mmcgr) [inline src/AstTransforms.cpp:320]: This explanation could be moved into the method comment in the header file.

**Comment by rdowavic** [comment]
alright cool i made the changes sorry abotu the delay

**Pull request merged**